students/jsward/session09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python
+
+
+class Donor:
+    def __init__(self, name, donations=()):
+        """
+        Initializes an instance of the Donor class
+
+        :param name: Donor's name as a string, or value that can be converted to a string
+        :param donations: Optional tuple of donations.  Donation values must be convertible to floats.
+        """
+        self.name = str(name)
+        self.donations = []
+        try:
+            for donation in list(donations):
+                self.add_donation(donation)
+        except TypeError:
+            self.add_donation(donations)
+
+    def __str__(self):
+        """
+        String representation of the donor object
+
+        :return: string
+        """
+        return self.name
+
+    @property
+    def stats(self):
+        """
+        Calculates donor statistics
+
+        :return: A list containing donor's name, total donation amount, number of donations and average donation amount
+        """
+        num_donations = len(self.donations)
+        total_donation = float(sum(self.donations))
+        if total_donation > 0:
+            average_donation = (float(total_donation / num_donations))
+        else:
+            average_donation = 0.00
+        return [self.name, total_donation, num_donations, average_donation]
+
+    @property
+    def thank_you_letter(self):
+        """
+        A thank you letter to the donor for their most recent donation.
+        If no donations have been made letter instead encourages making a first time donation.
+
+        :return: string
+        """
+        if len(self.donations) == 0:
+            letter_contents = "Dear {}, we look forward to your first donation!".format(self.name)
+        else:
+            letter_contents = "Dear {}, thank you for the very generous donation of ${:.2f}.  " \
+                              "Rest assured it will be put to good use helping those in need*\n" \
+                              "    *After deducting 98% for management overhead, of course ;).".format(self.name,
+                                                                                                   self.donations[-1])
+        return letter_contents
+
+    def add_donation(self, amount):
+        """
+        Add a donation from the donor
+
+        :param amount: The monetary value of the donation in USD.  Must be convertible to a float.
+        :return: None
+        """
+        if type(amount) is str:
+            amount = amount.strip('$ ').replace(',', '')
+        amount = float(amount)
+        if amount < 0:
+            raise ValueError
+        self.donations.append(amount)
+
+
+class DonorCollection:
+    def __init__(self):
+        """
+        Initializes an instance of the DonorCollection class
+        """
+        self.donors = []
```

Comment:
A list is a great start but a dict would have been a good option to use here. If you have a lot of donors in this list, you will have to iterate through the list every single time you want to find a donor. If you use a dict, that becomes much much faster.

Reply:
In class, we went over creating a normalized name. In practice, that would be the key used for the donor object in this dict.